src/kvstore/comm.h

```diff
@@ -635,6 +656,302 @@ class CommDevice : public Comm {
+                                GetNCCLType(src[i].dtype()),
+                                ncclSum,
+                                root,
+                                cur.comm,
+                                cur.stream););
+              }
+            }
+            ncclGroupEnd();
+          }
+        },
+        Context::CPU(),
+        const_vars,
+        {reduce.var()},
+        FnProperty::kCPUPrioritized,
+        priority,
+        PROFILER_MESSAGE("KVStoreReduce"));
+
+    return buf.merged;
+  }
+
+  void CommSync(const std::vector<const NDArray*>& dst,
+                int priority) override {
+    std::vector<Engine::VarHandle> const_vars;
+    std::vector<Engine::VarHandle> mutate_vars;
+    for (size_t i = 0; i < dst.size(); ++i) {
+        mutate_vars.push_back(dst[i]->var());
+    }
+    Engine::Get()->PushSync([this](RunContext rctx) {
+          for (auto cur : nccl_data_) {
+            CUDA_CALL(cudaSetDevice(cur.second.dev_id));
+            CUDA_CALL(cudaStreamSynchronize(cur.second.stream));
+          }
+        },
+        Context::CPU(),
+        const_vars,
+        mutate_vars,
+        FnProperty::kCPUPrioritized,
+        priority,
+        PROFILER_MESSAGE("KVStoreStreamSync"));
+  }
+
+  void CommSync(const std::vector<NDArray>& dst,
+                int priority) override {
+    std::vector<Engine::VarHandle> const_vars;
+    std::vector<Engine::VarHandle> mutate_vars;
+    for (size_t i = 0; i < dst.size(); ++i) {
+        mutate_vars.push_back(dst[i].var());
+    }
+    Engine::Get()->PushSync([this](RunContext rctx) {
+          for (auto cur : nccl_data_) {
+            CUDA_CALL(cudaSetDevice(cur.second.dev_id));
+            CUDA_CALL(cudaStreamSynchronize(cur.second.stream));
+          }
+        },
+        Context::CPU(),
+        const_vars,
+        mutate_vars,
+        FnProperty::kCPUPrioritized,
+        priority,
+        PROFILER_MESSAGE("KVStoreStreamSync"));
+  }
+
+  void BroadcastRowSparse(int key, const NDArray& src,
+                          const std::vector<std::pair<NDArray*, NDArray>>& dst,
+                          const bool use_copy,
+                          const int priority) override {
+    LOG(FATAL) << "NCCL kvstore does not support sparse storage type";
+  }
+
+  void Broadcast(int key, const NDArray& src,
+                 const std::vector<NDArray> dst, int priority) override {
+    if (!inited_) {
+      // copy to a random device first
+      int dev_id = key % dst.size();
+      CopyFromTo(src, dst[dev_id], priority);
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if (i != static_cast<size_t>(dev_id)) {
+          CopyFromTo(dst[dev_id], dst[i], priority);
+        }
+      }
+    } else {
+      auto& buf = merge_buf_[key];
+      int root = src.ctx().dev_id;
+      assert(root == buf.ctx().dev_id);
+      size_t root_id = -1;
+      for (size_t i = 0; i < dst.size(); ++i) {
+        if (dst[i].ctx().dev_id == root) {
+          root_id = i;
```

Comment:
could you please wrap such tasks in small wrapper functions so the code becomes more readable?

Reply:
I did for this snippet. Unfortunately the other small functions like this have slightly different elements so it's not as simple to wrap them and reuse. I also added some comments on what is being done.